### PR TITLE
Allow methods to be used for flexible searchable_fields

### PR DESCRIPTION
**Pull request by tiller1010**
### Summary
Allow methods to be used for flexible `$searchable_fields`. A method name can be added as a `$searchable_fields` key. The method should return an array of fields that can be used when using this search field in a GridFieldFilterHeader.

### Testing
- [ ] create a DataObject with `$searchable_fields`
- [ ] set a key to a method name that returns the appropriate search fields in an array (see comments in SearchContext)
- [ ] create records that use the searchable fields returned in your method
- [ ] confirm the search results will show any records that match at least one of the returned search fields

### Concerns
- It was necessary for me to define the `field` type in the searchable field options.

### CODE REVIEW

**Comment by emteknetnz** [comment]
Could you help me understand what this is supposed to do from a UI perspective?  Based on
```
                        public function getSearchableFirstName($queryName = '')
                        {
                            return [
                                'Customer.FirstName' => $queryName,
                                'ShippingAddress.FirstName' => $queryName,
                            ];
                        }
```
I'm assuming that when a user searches on FirstName it would now search on both the Customer.FirstName database field as well as the ShippingAddress.FirstName field?

**Comment by tiller1010** [comment]
@emteknetnz 
>I'm assuming that when a user searches on FirstName it would now search on both the Customer.FirstName database field as well as the ShippingAddress.FirstName field?

That's exactly right.

**Review by emteknetnz — CHANGES_REQUESTED** [review]
I'm finding the syntax of this confusing.  I don't think there's a need for a separate method as we're limited to what's in the database, and we cannot use dynamic content in this context i.e. filtering a DataList (as far as I'm aware, maybe I'm wrong)

I think putting everything in $searchable_fields would make more sense

```
private static $searchable_fields = [
    'CustomFirstName' => [
        'title' => 'First Name',
        'field' => TextField::class,
        'filter' => 'PartialMatchFilter',
        'match_any' => [
            'Customer.FirstName,
            'ShippingAddress.FirstName'
        ]
    ]
]
```

To merge this PR we'll also need unit test coverage added to this PR, as well as documentation about this additional functionality updated [here](https://github.com/silverstripe/silverstripe-framework/blob/4/docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md) rather than as an code comment as it is now

File: docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md [hunk 1]
```diff
@@ -154,6 +154,31 @@ class Player extends DataObject
 
 ```
 
+Use a single search field that matchs on multiple database fields with `'match_any'`
```
Comment (GuySartorelli) [inline docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md:157]:
```suggestion
Use a single search field that matches on multiple database fields with `'match_any'`
```

File: tests/php/ORM/Search/SearchContextTest/Address.php [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+use SilverStripe\ORM\DataObject;
+
+class Address extends DataObject implements TestOnly
+{
+    private static $table_name = 'SearchContextTest_Address';
+
+    private static $db = [
+        'FirstName' => 'Text'
+    ];
+}
```
Comment (michalkleiner) [inline tests/php/ORM/Search/SearchContextTest/Address.php:15]: Please keep an empty line at the end of the files, thanks

File: tests/php/ORM/Search/SearchContextTest/Customer.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+use SilverStripe\ORM\DataObject;
+
+class Customer extends DataObject implements TestOnly
+{
+    private static $table_name = 'SearchContextTest_Customer';
+
+    private static $db = [
+        'FirstName' => 'Text'
+    ];
+}
```
Comment (michalkleiner) [inline tests/php/ORM/Search/SearchContextTest/Customer.php:15]: Please keep a new line at the end of the files, thanks

File: tests/php/ORM/Search/SearchContextTest/Order.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            'field' => TextField::class,
+            'filter' => 'PartialMatchFilter',
+            'match_any' => [
+                // Searching with "First Name" will show Orders with matching Customer or Address names
+                'Customer.FirstName',
+                'ShippingAddress.FirstName',
+            ]
+        ]
+    ];
+}
```
Comment (michalkleiner) [inline tests/php/ORM/Search/SearchContextTest/Order.php:30]: Please keep a new line at the end of the files, thanks

**Review by michalkleiner — APPROVED** [review]
Looks ok from my perspective, thanks!

File: src/ORM/Search/SearchContext.php [hunk 5]
```diff
@@ -179,7 +180,34 @@ public function getQuery($searchParams, $sort = false, $limit = false, $existing
                 $filter->setModel($this->modelClass);
                 $filter->setValue($value);
                 if (!$filter->isEmpty()) {
-                    $query = $query->alterDataQuery([$filter, 'apply']);
+                    $modelObj = Injector::inst()->create($this->modelClass);
+                    if(isset($modelObj->searchableFields()[$key]['match_any'])) {
+                        $query = $query->alterDataQuery(function ($dataQuery) use ($modelObj, $key, $value) {
+                            $searchFields = $modelObj->searchableFields()[$key]['match_any'];
+                            $sqlSearchFields = [];
+                            foreach($searchFields as $dottedRelation){
```
Comment (emteknetnz) [inline src/ORM/Search/SearchContext.php:188]:
```suggestion
                            foreach ($searchFields as $dottedRelation){
```

File: src/ORM/Search/SearchContext.php [hunk 6]
```diff
@@ -179,7 +180,34 @@ public function getQuery($searchParams, $sort = false, $limit = false, $existing
...
+                                $relation = substr($dottedRelation, 0, strpos($dottedRelation, '.'));
+                                $relations = explode('.', $dottedRelation);
+                                $fieldName = array_pop($relations);
+
+                                // Apply join
```
Comment (emteknetnz) [inline src/ORM/Search/SearchContext.php:193]: I don't think we need these inline comments explaining the code or the line breaks above them

File: docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md [hunk 7]
```diff
@@ -154,6 +154,31 @@ class Player extends DataObject
...
+        'ShippingAddress' => Address::class,
+    ];
+
+    private static $searchable_fields = [
+        'CustomFirstName' => [
+            'title' => 'First Name',
+            'field' => TextField::class,
+            'filter' => 'PartialMatchFilter',
+            'match_any' => [
+                // Searching with "First Name" will show Orders with matching Customer or Address names
```
Comment (emteknetnz) [inline docs/en/02_Developer_Guides/00_Model/11_Scaffolding.md:173]:
```suggestion
                // Searching with the "First Name" field will show Orders matching either Customer.FirstName or ShippingAddress.FirstName
```

**Review by emteknetnz — CHANGES_REQUESTED** [review]
Great work.

I've tested this locally to confirm it works as expected

Some very minor changes then I'm happy to merge

There's some PHPCS linting to fix as well https://app.travis-ci.com/github/silverstripe/silverstripe-framework/jobs/558865008

**Review by emteknetnz — APPROVED** [review]

**Pull request merged**